src/parser.rs

```diff
@@ -551,175 +557,199 @@ pub enum SegmentTag {
-        ,
-        |pairs: Vec<(String, String)>| {
-            pairs.into_iter().collect()
+pub fn key_value_pairs(i: &[u8]) -> IResult<&[u8], HashMap<String, String>> {
+    fold_many0(
+        preceded(space0, key_value_pair),
+        HashMap::new,
+        |mut acc: HashMap<_, _>, (left, right)| {
+            acc.insert(left, right);
+            acc
+        },
+    )(i)
+}
+
+pub fn key_value_pair(i: &[u8]) -> IResult<&[u8], (String, String)> {
+    map(
+        tuple((
+            peek(none_of("\r\n")),
+            map_res(take_until("="), from_utf8_slice),
+            char('='),
+            alt((quoted, unquoted)),
+            opt(char(',')),
+        )),
+        |(_, left, _, right, _)| (left, right),
+    )(i)
+}
+
+pub fn quoted(i: &[u8]) -> IResult<&[u8], String> {
+    delimited(
+        char('\"'),
+        map_res(is_not("\""), from_utf8_slice),
+        char('\"'),
+    )(i)
+}
+
+pub fn unquoted(i: &[u8]) -> IResult<&[u8], String> {
+    map_res(is_not(",\r\n"), from_utf8_slice)(i)
+}
+
+pub fn consume_line(i: &[u8]) -> IResult<&[u8], String> {
+    map(
+        pair(map_res(not_line_ending, from_utf8_slice), opt(line_ending)),
+        |(line, _)| line,
+    )(i)
+}
+
+pub fn number(i: &[u8]) -> IResult<&[u8], i32> {
+    map_res(take_while1(is_digit), |s| {
+        // Can't fail because we validated it above already
+        let s = str::from_utf8(s).unwrap();
+        str::parse::<i32>(s)
+    })(i)
+}
+
+pub fn byte_range_val(i: &[u8]) -> IResult<&[u8], ByteRange> {
+    map(pair(number, opt(preceded(char('@'), number))), |(n, o)| {
+        ByteRange {
+            length: n,
+            offset: o,
         }
-    )
-);
-
-named!(pub key_value_pair(&[u8]) -> (String, String),
-    do_parse!(
-          peek!(none_of!("\r\n"))
-       >> left: map_res!(take_until!("="), from_utf8_slice)
-       >> take!(1)
-       >> right: alt!(quoted | unquoted)
-       >> opt!(char!(','))
-       >>
-       (left, right)
-    )
-);
-
-named!(pub quoted<String>,
-    delimited!(char!('\"'), map_res!(is_not!("\""), from_utf8_slice), char!('\"'))
-);
-
-named!(pub unquoted<String>,
-    map_res!(is_not!(",\r\n"), from_utf8_slice)
-);
-
-named!(pub consume_line<String>,
-    do_parse!(
-        line: map_res!(not_line_ending, from_utf8_slice)
-        >> opt!(line_ending)
-        >> (line)
-    )
-);
-
-named!(pub number<i32>,
-    map_res!(map_res!(digit1, str::from_utf8), str::FromStr::from_str)
-);
-
-named!(pub byte_range_val<ByteRange>,
-    do_parse!(
-          n: number
-       >> o: opt!(do_parse!(char!('@') >> n:number >> (n) )) >>
-        (ByteRange { length: n, offset: o })
-    )
-);
-
-named!(pub float<f32>,
-    do_parse!(
-               left: map_res!(digit1, str::from_utf8)
-       >> right_opt: opt!(do_parse!(char!('.') >> d:map_res!(digit1, str::from_utf8) >> (d) ))
-       >>
-        (
-        match right_opt {
+    })(i)
+}
+
+pub fn float(i: &[u8]) -> IResult<&[u8], f32> {
```

Comment:
This now works without going through intermediate, temporary `String` allocations

Reply:
Nice 👍